Add user and password parameters

avlitman opened the pull request:
We want to expose the uid and pwd parameters in our role,
to allow basic auth on the elasticsearch server.

The metrics update can be seen here: https://gerrit.ovirt.org/#/c/ovirt-engine-metrics/+/116098/

Bug-Url: https://bugzilla.redhat.com/1990490
Signed-off-by: Aviv Litman <alitman@redhat.com>

richm commented:
[citest commit:cfb15aba2c94ba88f63f6b84f5231d6fc201f5db]

nhosoi commented on roles/rsyslog/templates/output_elasticsearch.j2:
```diff
@@ -86,6 +86,14 @@ ruleset(name="{{ item.name }}") {
+{%   if item.pwd is defined %}
+{%     set __pwd = item.pwd %}
+{%   endif %}
+            tls.uid="{{ __uid }}"
```
@avlitman, I have 2 questions.
1) I cannot find the parameters `tls.uid` and `tls.pwd` in the rsyslog omelasiticsearch doc:
https://www.rsyslog.com/doc/v8-stable/configuration/modules/omelasticsearch.html
I see `uid` and `pwd` in the doc. Could they be a typo?

2) If so, please note that `uid` and `pwd` are for the "basic authentication" (not over `tls`).
The section `[2-6-4] Set built-in user password` in https://www.elastic.co/blog/configuring-ssl-tls-and-https-to-secure-elasticsearch-kibana-beats-and-logstash for setting the password of these built-in users.
```
elastic,apm_system,kibana,logstash_system,beats_system,remote_monitoring_user
```
Are you planning to let rsyslog omelasticsearch plugin to access elasticsearch using one of these built-in users with the basic authentication?

richm replied:
The rsyslog omelasticsearch plugin uses `uid` and `pwd` - https://github.com/rsyslog/rsyslog/blob/master/plugins/omelasticsearch/omelasticsearch.c#L190

nhosoi replied:
Thanks, @richm. I meant "Could `tls.uid` and `tls.pwd` be a typo?" :p
(I grepped `tls.uid` and `tls.pwd` in the rsyslog omelasticsearch source code to find they are not in it. :)

richm replied:
> Thanks, @richm. I meant "Could `tls.uid` and `tls.pwd` be a typo?" :p
> (I grepped `tls.uid` and `tls.pwd` in the rsyslog omelasticsearch source code to find they are not in it. :)

Yes, it is a typo - it should be `uid` and `pwd`

avlitman replied:
Sorry, it's a typo.

avlitman replied:
@nhosoi 
Are you planning to let rsyslog omelasticsearch plugin to access elasticsearch using one of these built-in users with the basic authentication?

Is there a reason not to allow that ?

richm replied:
omelasticsearch will use TLS if you specify `usehttps: true` - including if the user specifies `uid` and `pwd` - omelasticsearch will make an HTTPS connection and will send `uid` and `pwd` over an encrypted connection.  It doesn't matter if you specify `ca_cert`, `cert`, or `private_key` - you only have to specify `usehttps: true`.  Of course, if you don't specify `ca_cert` then you will probably get some sort of TLS error, unless you also specify `allowUnsignedCerts: true` and also somehow defeat host verification checking.
If the user specifies all of `uid`, `pwd`,  `cert`, and `private_key`, then all of them will be sent to Elasticsearch, and it will be up to the Elasticsearch configuration as to which authentication method it will accept.  I believe SearchGuard allows you to say "use client cert if provided, otherwise, fallback to username/password authentication".

nhosoi replied:
Thanks, @richm. That means there's no further security issues other than storing the sensitive info in the conf file and the ansible log, is there? If that's the case, as long as the "risk" is known to the team/users, I have no more objection. Thanks.

richm replied:
> Thanks, @richm. That means there's no further security issues other than storing the sensitive info in the conf file and the ansible log, is there?

Right - and ansible vault can be used to protect sensitive information in the inventory/playbook.

> If that's the case, as long as the "risk" is known to the team/users, I have no more objection. Thanks.

nhosoi replied:
> Right - and ansible vault can be used to protect sensitive information in the inventory/playbook.

If ansible vault is used, will these values not appear in these config files in /etc/rsyslog.d? Or we do not care? Please note that current mode of /etc/rsyslog.d is 0755 and the one of the config files is 0644 (unless RHV/oVirt changes them).
```
31-output-elasticsearch-elasticsearch_output_ops.conf:            uid="testuser0"
31-output-elasticsearch-elasticsearch_output_ops.conf-            pwd="testpassword0"
```

avlitman replied:
I think the value (only pwd need encrypting) will appear in the 31-* files as asterisks: pwd: "******"

avlitman replied:
> Right - and ansible vault can be used to protect sensitive information in the inventory/playbook.

@richm Richard, do you familiar with the ansible vault, and can help me with that?

Thanks

nhosoi replied:
> I think the value (only pwd need encrypting) will appear in the 31-* files as asterisks: pwd: "******"

Sorry, I don't understand... To decrypt an encrypted value (or a file), we need `ansible-vault decrypt` or run ansible with the vault password, don't we? Once rsyslog config files are generated, there's no room for ansible to provide such a service? I thought `ansible-vault` protects sensitive data in the playbook and inventory file, but not in the generated files... Am I missing something?

mwperina replied:
Unless I'm missing something ansible vault can be used to hide password only during ansible playbook execution. But AFAIK we are passing parameters to logging role using facts, so as soon as you add password to the facts its value will be revealed in the log and all password protection is gone. The only option I see is to create a variable in logging role which can be used to pass the password and in this case we could use an ansible vault to protect password value during execution. Am I right @mnecas ?

Regarding storing username password in /etc/rsyslog.d do we really need the relevant config file to have 644 permissions set? Couldn't we use 600 or 640 so only user under which rsyslogd is running can read the password?

nhosoi replied:
@mwperina,
> Regarding storing username password in /etc/rsyslog.d do we really need the relevant config file to have 644 permissions set? Couldn't we use 600 or 640 so only user under which rsyslogd is running can read the password?

We changed the mode to 0600 here in the latest commits.
https://github.com/linux-system-roles/logging/pull/236/files#diff-b7d4f0de7f6ad2b775fec097d0fc62f356c66dafe1c18d397c12d0da75f2362dR27
Thanks!

nhosoi commented:
@avlitman, one more request...

The tests_ovirt_elasticsearch.yml test fails if `uid` and `pwd` are not given. Please move the lines inside `is defined`.

```
diff --git a/roles/rsyslog/templates/output_elasticsearch.j2 b/roles/rsyslog/templates/output_elasticsearch.j2
index 7ff1f9c..9fe3baf 100644
--- a/roles/rsyslog/templates/output_elasticsearch.j2
+++ b/roles/rsyslog/templates/output_elasticsearch.j2
@@ -88,12 +88,12 @@ ruleset(name="{{ item.name }}") {
             tls.myprivkey="{{ __myprivkey }}"
 {%   if item.uid is defined %}
 {%     set __uid = item.uid %}
+            uid="{{ __uid }}"
 {%   endif %}
 {%   if item.pwd is defined %}
 {%     set __pwd = item.pwd %}
-{%   endif %}
-            uid="{{ __uid }}"
             pwd="{{ __pwd }}"
+{%   endif %}
 {% endif %}
         )
     }
```

nhosoi commented:
> The ownership/permissions on the 31-output-elasticsearch-elasticsearch_*.conf files should be as tight as possible - use mode 0600 if possible so that only `root` or `rsyslog` can read the file that contains the cleartext password.

We are setting the ownership and the permission here.  Would the change be made only on `31-output-elasticsearch-elasitcsearch-*.conf`? (I don't think we are supposed to change the mode of all the config files to 0600...) And would it be done in this task or in the RHV/ovirt integration?
https://github.com/linux-system-roles/logging/blob/master/roles/rsyslog/tasks/main.yml#L151-L153

nhosoi commented:
@avlitman, 

Sorry, my request was incomplete. We don't have to set `item.uid` to `__uid`, but we can just use  `item.uid` as ` uid="{{ item.uid }}"`. (The same for `pwd`, too.)

Prev:
```
 {%   if item.uid is defined %}
 {%     set __uid = item.uid %}
            uid="{{ __uid }}"
 {%   endif %}
```
New:
```
 {%   if item.uid is defined %}
            uid="{{ item.uid }}"
 {%   endif %}
```
And could you please rebase this pr against the master branch before pushing to solve the CI test failures? It looks to me github is confused with some already merged commits...

richm commented:
> > The ownership/permissions on the 31-output-elasticsearch-elasticsearch_*.conf files should be as tight as possible - use mode 0600 if possible so that only `root` or `rsyslog` can read the file that contains the cleartext password.
> 
> We are setting the ownership and the permission here. Would the change be made only on `31-output-elasticsearch-elasitcsearch-*.conf`? (I don't think we are supposed to change the mode of all the config files to 0600...)

As long as we can guarantee that files containing cleartext passwords, or other secrets, are only readable by root or rsyslog, that is fine.  If we can do that without changing the mode of all config files to 0600 that is fine.

> And would it be done in this task or in the RHV/ovirt integration?

It would be done in the task that creates the file.

> https://github.com/linux-system-roles/logging/blob/master/roles/rsyslog/tasks/main.yml#L151-L153

nhosoi commented:
> > > The ownership/permissions on the 31-output-elasticsearch-elasticsearch_*.conf files should be as tight as possible - use mode 0600 if possible so that only `root` or `rsyslog` can read the file that contains the cleartext password.
> > 
> > We are setting the ownership and the permission here. Would the change be made only on `31-output-elasticsearch-elasitcsearch-*.conf`? (I don't think we are supposed to change the mode of all the config files to 0600...)
> 
> As long as we can guarantee that files containing cleartext passwords, or other secrets, are only readable by root or rsyslog, that is fine. If we can do that without changing the mode of all config files to 0600 that is fine.
> 
> > And would it be done in this task or in the RHV/ovirt integration?
> 
> It would be done in the task that creates the file.

@avlitman, I opened [a pr](https://github.com/avlitman/logging/pull/2) against your branch avlitman:add_user_and_password_ELK.
- 31-output-elasticsearch-elasticsearch_output*.conf: set mode to 0600.
- output_elasticsearch.j2: eliminate unnecessary __uid and __pwd setting.

Could you please take a look? (It contains the change described in https://github.com/linux-system-roles/logging/pull/236#issuecomment-896156991, too.) Thanks.

avlitman commented:
> > > > The ownership/permissions on the 31-output-elasticsearch-elasticsearch_*.conf files should be as tight as possible - use mode 0600 if possible so that only `root` or `rsyslog` can read the file that contains the cleartext password.
> > > 
> > > 
> > > We are setting the ownership and the permission here. Would the change be made only on `31-output-elasticsearch-elasitcsearch-*.conf`? (I don't think we are supposed to change the mode of all the config files to 0600...)
> > 
> > 
> > As long as we can guarantee that files containing cleartext passwords, or other secrets, are only readable by root or rsyslog, that is fine. If we can do that without changing the mode of all config files to 0600 that is fine.
> > > And would it be done in this task or in the RHV/ovirt integration?
> > 
> > 
> > It would be done in the task that creates the file.
> 
> @avlitman, I opened [a pr](https://github.com/avlitman/logging/pull/2) against your branch avlitman:add_user_and_password_ELK.
> 
> * 31-output-elasticsearch-elasticsearch_output*.conf: set mode to 0600.
> * output_elasticsearch.j2: eliminate unnecessary __uid and __pwd setting.
> 
> Could you please take a look? (It contains the change described in [#236 (comment)](https://github.com/linux-system-roles/logging/pull/236#issuecomment-896156991), too.)
> 
> Pushde another commit to [avlitman#2](https://github.com/avlitman/logging/pull/2).
> [avlitman@72b0763](https://github.com/avlitman/logging/commit/72b0763b1a1d769a23b0ad7b669eaf2270250f93)
> The change prevents to output the sensitive task results by setting `no_log: true`. But it could make a troubleshooting harder.

Thank you Noriko, 
I will test it in my environment asap.

avlitman commented:
> @avlitman,
> 
> Sorry, my request was incomplete. We don't have to set `item.uid` to `__uid`, but we can just use `item.uid` as ` uid="{{ item.uid }}"`. (The same for `pwd`, too.)
> 
> Prev:
> 
> ```
>  {%   if item.uid is defined %}
>  {%     set __uid = item.uid %}
>             uid="{{ __uid }}"
>  {%   endif %}
> ```
> 
> New:
> 
> ```
>  {%   if item.uid is defined %}
>             uid="{{ item.uid }}"
>  {%   endif %}
> ```
> 
> And could you please rebase this pr against the master branch before pushing to solve the CI test failures? It looks to me github is confused with some already merged commits...

Rebase now? or later?

nhosoi commented:
> Rebase now? or later?

Could you rebase now? I'd like to see if it solves the CI tests failures. Thanks, @avlitman!

nhosoi commented:
[citest bad]

nhosoi commented:
@avlitman, @richm, let's decide the fate of this pr.
- In this pr, the elasticsearch output accepts the user and password.
- The rsyslog omelasticsearch config files are explicitly owned by root and their mode is 0600.
- And we set `no_log` to about a dozen tasks not to output which could output the password in the log. They are quite generic tasks and no matter how many verbose option is set (e.g., -vvvv), the output is terse like this.
```
TASK [/WORKDIR/dist-git-logging-add_user_and_password_ELK-YuUTaD/roles/rsyslog : Generate role configuration files in rsyslog.d] ***
skipping: [sut] => {
    "censored": "the output has been hidden due to the fact that 'no_log: true' was specified for this result", 
    "changed": false
}
```
It's hard to learn which input/output was processed and in this case, why it was skipped.
It is true for all the use cases with or without the elasticsearch output.
I'm a little worried that it might make the non-elasticsearch users' life harder as they will get less info when debugging their playbooks. To workaround this, we may be able to make a copy from each of the task set to no_log for the elasticsearch output. And set no_log to true just for the elasticsearch with password in the copy. (It's getting uglier though...)

richm commented:
>  It is true for all the use cases with or without the elasticsearch output.

yeah - unfortunately the `no_log` feature has no granularity - it's either all or nothing - ideally we would have a way to mask only the `pwd` value, but since that is nested within the other parameters, it is not possible.

Another idea - we make the elasticsearch password a separate top level parameter e.g.
`logging_elasticsearch_password: some value`
Pros:
- makes it easy to use no_log?
- makes it easier to use vault
Cons:
- breaks the model we currently use to pass in parameters
- would require some rewrite of ovirt code

nhosoi commented:
> > It is true for all the use cases with or without the elasticsearch output.
> 
> yeah - unfortunately the `no_log` feature has no granularity - it's either all or nothing - ideally we would have a way to mask only the `pwd` value, but since that is nested within the other parameters, it is not possible.
> 
> Another idea - we make the elasticsearch password a separate top level parameter e.g.
> `logging_elasticsearch_password: some value`
> Pros:
> 
>     * makes it easy to use no_log?
> 
>     * makes it easier to use vault
>       Cons:
> 
>     * breaks the model we currently use to pass in parameters
> 
>     * would require some rewrite of ovirt code

I quickly tried this option, and found the password is shown only at two places in the output log! But one of them is this task we'd like to make it fully logged...
```
TASK [/path/to/linux-system-roles/logging/roles/rsyslog : Generate role configuration files in rsyslog.d] ***
task path: /path/to/linux-system-roles/logging/roles/rsyslog/tasks/deploy.yml:13
```
Next, I'm trying to combine the two ideas, 1) introducing a variable logging_elasticsearch_password to reduce the tasks where the password could be logged and 2) making a copy of the task "Generate role configuration files in rsyslog.d" with no_log.

nhosoi commented:
@avlitman, this pr was merged from the master branch. It should be "rebased". It'd ensure your changes are made on top of the master branch. Otherwise, we cannot merge this pr into the master branch. To fix it, please do the followings.

Make sure you have `[remote avlitman]` in your `logging/.git/config`.
If you do not have one, run the following commands.
```
git checkout master
git pull
git remote add avlitman https://github.com/avlitman/logging
git fetch avlitman
```
Then, delete the accidentally merged commits (plus, I'd like you to remove my previous no_log commit), rebase against the master branch, and push your changes to your branch.
```
git checkout add_user_and_password_ELK
git reset --hard a359d33e7e2f9eb21afc0bc45a814a0a0ef93b8b
git rebase master
git push --force avlitman add_user_and_password_ELK
```
Once it's done, I'm going to open another pr against your add_user_and_password_ELK branch. Thanks.

avlitman commented:
> @avlitman, this pr was merged from the master branch. It should be "rebased". It'd ensure your changes are made on top of the master branch. Otherwise, we cannot merge this pr into the master branch. To fix it, please do the followings.
> 
> Make sure you have `[remote avlitman]` in your `logging/.git/config`.
> If you do not have one, run the following commands.
> 
> ```
> git checkout master
> git pull
> git remote add avlitman https://github.com/avlitman/logging
> git fetch avlitman
> ```
> 
> Then, delete the accidentally merged commits (plus, I'd like you to remove my previous no_log commit), rebase against the master branch, and push your changes to your branch.
> 
> ```
> git checkout add_user_and_password_ELK
> git reset --hard a359d33e7e2f9eb21afc0bc45a814a0a0ef93b8b
> git rebase master
> git push --force avlitman add_user_and_password_ELK
> ```
> 
> Once it's done, I'm going to open another pr against your add_user_and_password_ELK branch. Thanks.

Done. 
Thank you!

nhosoi commented:
Hi @avlitman, I've opened another [pr](https://github.com/avlitman/logging/pull/5) against your branch add_user_and_password_ELK. This is a simple workaround made from 2 changes: introducing an elaticsearch password var logging_elasticsearch_password and deploy_nolog.yml which does not log. As you see in the pr/5, the elasticsearch output task includes the task file deploy_nolog.yml.

Basically, this is an implementation of @richm's idea in https://github.com/linux-system-roles/logging/pull/236#issuecomment-902041907. As mentioned in the list of `Cons` the ovirt needs to remove the pwd parameter and add the logging_elasticsearch_password var. But it's much simpler than asking the customers to set up vault, IMHO.

avlitman commented:
> Hi @avlitman, I've opened another [pr](https://github.com/avlitman/logging/pull/5) against your branch add_user_and_password_ELK. This is a simple workaround made from 2 changes: introducing an elaticsearch password var logging_elasticsearch_password and deploy_nolog.yml which does not log. As you see in the pr/5, the elasticsearch output task includes the task file deploy_nolog.yml.
> 
> Basically, this is an implementation of @richm's idea in [#236 (comment)](https://github.com/linux-system-roles/logging/pull/236#issuecomment-902041907). As mentioned in the list of `Cons` the ovirt needs to remove the pwd parameter and add the logging_elasticsearch_password var. But it's much simpler than asking the customers to set up vault, IMHO.

ack.
In addition I want to consult on an approach I saw:
Add a debug task with regex_replace like mentioned here: https://serverfault.com/questions/681832/how-can-i-stop-ansible-from-writing-passwords-to-the-logfiles
to also get the output.

nhosoi commented:
> > Hi @avlitman, I've opened another [pr](https://github.com/avlitman/logging/pull/5) against your branch add_user_and_password_ELK. This is a simple workaround made from 2 changes: introducing an elaticsearch password var logging_elasticsearch_password and deploy_nolog.yml which does not log. As you see in the pr/5, the elasticsearch output task includes the task file deploy_nolog.yml.
> > Basically, this is an implementation of @richm's idea in [#236 (comment)](https://github.com/linux-system-roles/logging/pull/236#issuecomment-902041907). As mentioned in the list of `Cons` the ovirt needs to remove the pwd parameter and add the logging_elasticsearch_password var. But it's much simpler than asking the customers to set up vault, IMHO.
> 
> ack.
> @nhosoi @richm
> In addition I want to consult about an approach I saw:
> Add a debug task with regex_replace like mentioned here: https://serverfault.com/questions/681832/how-can-i-stop-ansible-from-writing-passwords-to-the-logfiles
> to also get the output.
> I can try to implement

Sorry, I don't understand your proposal. With the pr/5 (https://github.com/avlitman/logging/pull/5) opened against this pr, the `password0` used in the test does not appear in the ansible output nor in the system log (in `journalctl` output nor in `/var/log/messages`).
https://github.com/avlitman/logging/pull/5/files#diff-9b668b7a142c917d288498e9fb3f61a87f4bdc58cb4c94436781c46eca133170R181

Could you please tell me what you are trying to solve with your proposal?

avlitman commented:
> > > Hi @avlitman, I've opened another [pr](https://github.com/avlitman/logging/pull/5) against your branch add_user_and_password_ELK. This is a simple workaround made from 2 changes: introducing an elaticsearch password var logging_elasticsearch_password and deploy_nolog.yml which does not log. As you see in the pr/5, the elasticsearch output task includes the task file deploy_nolog.yml.
> > > Basically, this is an implementation of @richm's idea in [#236 (comment)](https://github.com/linux-system-roles/logging/pull/236#issuecomment-902041907). As mentioned in the list of `Cons` the ovirt needs to remove the pwd parameter and add the logging_elasticsearch_password var. But it's much simpler than asking the customers to set up vault, IMHO.
> > 
> > 
> > ack.
> > @nhosoi @richm
> > In addition I want to consult about an approach I saw:
> > Add a debug task with regex_replace like mentioned here: https://serverfault.com/questions/681832/how-can-i-stop-ansible-from-writing-passwords-to-the-logfiles
> > to also get the output.
> > I can try to implement
> 
> Sorry, I don't understand your proposal. With the pr/5 ([avlitman#5](https://github.com/avlitman/logging/pull/5)) opened against this pr, the `password0` used in the test does not appear in the ansible output nor in the system log (in `journalctl` output nor in `/var/log/messages`).
> https://github.com/avlitman/logging/pull/5/files#diff-9b668b7a142c917d288498e9fb3f61a87f4bdc58cb4c94436781c46eca133170R181
> 
> Could you please tell me what you are trying to solve with your proposal?

If I understand correctly, the pr hide all variables from the output.
I'm trying to hide just the password.
In addition, maybe we can create the no log tasks only when uid and password defined ?

avlitman commented:
> Hi @avlitman, I've opened another [pr](https://github.com/avlitman/logging/pull/5) against your branch add_user_and_password_ELK. This is a simple workaround made from 2 changes: introducing an elaticsearch password var logging_elasticsearch_password and deploy_nolog.yml which does not log. As you see in the pr/5, the elasticsearch output task includes the task file deploy_nolog.yml.
> 
> Basically, this is an implementation of @richm's idea in [#236 (comment)](https://github.com/linux-system-roles/logging/pull/236#issuecomment-902041907). As mentioned in the list of `Cons` the ovirt needs to remove the pwd parameter and add the logging_elasticsearch_password var. But it's much simpler than asking the customers to set up vault, IMHO.

Hi @nhosoi,
I tested the [metrics patch](https://gerrit.ovirt.org/#/c/ovirt-engine-metrics/+/116098/4/roles/oVirt.logging/tasks/main.yml)+[pr236](https://github.com/linux-system-roles/logging/pull/236)+[this pr](https://github.com/avlitman/logging/pull/5). 

**can't see the password in the logs of tasks:**
Generate role configuration files in rsyslog.d, 
Remove role config files from rsyslog.d

**can see the password in the logs of tasks:**
Copy ca_cert on the control host to the specified path on the target host,
Copy cert on the control host to the specified path on the target host
Copy key on the control host to the specified path on the target host
Check certs - tls is true, but triplets are not given,
Check certs - key/certs data are provided, but tls is false

**The password shown in the 31 output files as:** 
pwd="the-password"

nhosoi commented:
Thanks for testing, @avlitman ! I'm going to update my pr/5...

nhosoi commented:
Hmmm, I cannot reproduce the "**password is visible in tasks**" part.

I'm wondering you might not have eliminated `pwd: <<password_string>>` from the individual elasticsearch output and moved the password to the global `logging_elasticsearch_password: <<password_string>>`?

The snippet of my test playbook looks like this. With this playbook, `the-real-password` is not observed in the ansible log.
```
  tasks:
    - name: <<TASK NAME>>
      vars:
        logging_elasticsearch_password: the-real-password
        logging_outputs:
          - name: elasticsearch_output_ops
            type: elasticsearch
            server_host: logging-es-ops
            server_port: 9200
            <<snip>>
            cert: "/etc/rsyslog.d/es-cert.pem"
            private_key: "/etc/rsyslog.d/es-key.pem"
            uid: <<user-id>>
            <<snip>>
```
I experimentally tested with `pwd: fake-password` added after `uid: <<user-id>>`, then `fake-password` appeared in the ansible log output, but not `the-real-password`. Please be careful, this https://github.com/avlitman/logging/pull/5 requires the change in the ovirt playbook, too.
See also the comment from @richm:
https://github.com/linux-system-roles/logging/pull/236#issuecomment-902041907
Cons:
- breaks the model we currently use to pass in parameters
- would require some rewrite of ovirt code

Plus, regarding **The password shown in the 31 output files as ...**, we agreed there is no way to encrypt the password in the rsyslog config files, right? Instead, we force to set the mode of the files to 0600, while the other files' mode is 0644.
```
-rw-------. 1 root root 2029 Aug 24 15:42 /etc/rsyslog.d/31-output-elasticsearch-elasticsearch_output.conf
-rw-------. 1 root root 2073 Aug 24 15:42 /etc/rsyslog.d/31-output-elasticsearch-elasticsearch_output_ops.conf
```

avlitman commented:
The only change I did is to set the the new var, you can see the change in [metrics patch](https://gerrit.ovirt.org/#/c/ovirt-engine-metrics/+/116098/5/roles/oVirt.logging/tasks/main.yml).
then my steps:
1. # ansible-vault create /etc/ovirt-engine-metrics/config.yml.d/secret.yml 
New Vault password: 
Confirm New Vault password: 

2. # ansible-vault edit /etc/ovirt-engine-metrics/config.yml.d/secret.yml
rsyslog_elasticsearch_pwd: "the-real-password"

3. run the playbook with --ask-vault-pass
# /usr/share/ovirt-engine-metrics/setup/ansible/configure_ovirt_machines_for_metrics.sh --ask-vault-pass

password is not visible in tasks: -> make sense since these are the tasks you add the no log option
Generate role configuration files in rsyslog.d,
Remove role config files from rsyslog.d

password is visible in tasks: -> make sense since these are the tasks you did not add the no log option
Copy ca_cert on the control host to the specified path on the target host,
Copy cert on the control host to the specified path on the target host
Copy key on the control host to the specified path on the target host
Check certs - tls is true, but triplets are not given,
Check certs - key/certs data are provided, but tls is false

nhosoi commented:
> The only change I did is to set the the new var, you can see the change in [metrics patch](https://gerrit.ovirt.org/#/c/ovirt-engine-metrics/+/116098/5/roles/oVirt.logging/tasks/main.yml): maybe it's not correct?

The location of `logging_elasticsearch_password` should be at the same level of `logging_outputs` (it should **not** be in each elasticsearch output config. I.e., there is only one `logging_elasticsearch_password` shared among all the elasticsearch configs (if there are multiple of them as in the ovirt playbook). Thanks!

avlitman commented:
> > The only change I did is to set the the new var, you can see the change in [metrics patch](https://gerrit.ovirt.org/#/c/ovirt-engine-metrics/+/116098/5/roles/oVirt.logging/tasks/main.yml): maybe it's not correct?
> 
> The location of `logging_elasticsearch_password` should be at the same level of `logging_outputs` (it should **not** be in each elasticsearch output config. I.e., there is only one `logging_elasticsearch_password` shared among all the elasticsearch configs (if there are multiple of them as in the ovirt playbook). Thanks!

Thank you!
I will fix and update

avlitman commented:
HI @nhosoi 
I fixed the metrics [patch](https://gerrit.ovirt.org/#/c/ovirt-engine-metrics/+/116098/).
Now I can't see the password in the logs in all flows.
Flows:
1. password is defiend in a secret.yml file, and tls is true: uid and password is in the 31 output files.
2. password is not defiend and no secret.yml file, and tls is true: uid  is in the 31 output files.
3. password is defiend in a secret.yml file, and tls is false: uid and password is not in the 31 output files.
4. password is not defiend and no secret.yml file, and tls is false: **uid** and password is not in the 31 output files.
5. password is not defiend and no secret.yml file, tls is false, and **config.yml contain the uid var**: **uid** and password is not in the 31 output files.

I asked Gui if it's possible to use password and user name whem tls is false,
If yes we need to change roles/rsyslog/templates/output_elasticsearch.j2 file.

other than that everything lgtm.

nhosoi commented:
Thanks for the complete test results, @avlitman. Great news. And I agree that `uid` and `pwd` should be available outside of `tls`.

I have one very minor comment in your metrics patch.

> I fixed the metrics patch.
```
139    logging_elasticsearch_password: "{{ logging_elasticsearch_password|d([]) }}"
```
Since logging_elasticsearch_password is a string (not a list), the default should be `d()` instead of `d([])`?

avlitman commented:
> And once you are ready to merge, could you please run
> 
> ```
> git rebase -i da5766cd40813c0f36c56972c1155f1a1fdbe5cf
> ```
> 
> and merge multiple commits into one? Thanks!

Thank you,
I merge pr/5 and multiple commits into one.
not sure why tox / python (3.8) is failing

avlitman commented:
[citest bad]

richm commented:
>  not sure why tox / python (3.8) is failing

molecule tests are currently broken

nhosoi reviewed (approved):
lgtm

sradco commented:
@nhosoi @avlitman Sorry for the late comment. I don't think we should always set use no_log to true for Elasticsearch output , only if the user has provided a password.

nhosoi commented:
> @nhosoi @avlitman Sorry for the late comment. I don't think we should always set use no_log to true for Elasticsearch output , only if the user has provided a password.

@sradco, thanks. I'm working on it now. Plus, we need to have updated the README.md.

avlitman commented:
> @sradco, @avlitman, I've submitted [avlitman#6](https://github.com/avlitman/logging/pull/6). If you could review it, I'd appreciate it.
 
 Thank you so much!
 Of course I will review and update ASAP.

avlitman commented:
> Of course I will review and update ASAP.

I reviewed  [avlitman#6](https://github.com/avlitman/logging/pull/6):
The elasticsearch output tasks behaves as expected :
no password-> no_log version of deploy.yml is not called, and logs are present.
password empty-> no_log version of deploy.yml is not called and logs are present.
password defined-> The no_log version of deploy.yml is called, and did not see the password string. 

I merged [avlitman#6](https://github.com/avlitman/logging/pull/6).

nhosoi commented:
> @nhosoi Let ne know if need to squash commits.

Thanks for reviewing & testing, @avlitman! Yes, please squash commits into one. Then, I'm going to merge.

avlitman commented:
> > @nhosoi Let ne know if need to squash commits.
> 
> Thanks for reviewing & testing, @avlitman! Yes, please squash commits into one. Then, I'm going to merge.

Done, and also update to latest master.
@nhosoi Thank you so much for all the help, much appreciated!

nhosoi commented:
Thank you for squashing, @avlitman. But there's something strainge...

In the commits page https://github.com/linux-system-roles/logging/pull/236/commits, there are two commits.
- Add user and password in elasticsearch output
- Merge branch 'linux-system-roles:master' into add_user_and_password_ELK

The second commit is reverting the changes made on roles/rsyslog/tasks/set_certs.yml and tests/tests_files_elasticsearch.yml. Not sure what operation made it occur, but the second commit should not be there.

If you do `git log` in your branch `add_user_and_password_ELK`, the top part would look like this.
```
commit 687ddc95e91dc7a0a8ca7d4113dfc83544099b78 (HEAD -> pr/236, origin/pr/236)
Merge: 59e7de1 5742b01
Author: avlitman <64130977+avlitman@users.noreply.github.com>
Date:   Tue Aug 31 20:18:34 2021 +0300

    Merge branch 'linux-system-roles:master' into add_user_and_password_ELK

commit 59e7de1e975e907d163dbd1586135f1618b2234d (tmp)
Author: Aviv Litman <alitman@redhat.com>
Date:   Thu Aug 5 16:24:51 2021 +0300

    Add user and password in elasticsearch output
```
Could you delete the top commit and push it as follows?
```
git reset --hard 59e7de1e975e907d163dbd1586135f1618b2234d
git push --force avlitman add_user_and_password_ELK
```

avlitman commented:
> Thank you for squashing, @avlitman. But there's something strainge...
> 
> In the commits page https://github.com/linux-system-roles/logging/pull/236/commits, there are two commits.
> 
> * Add user and password in elasticsearch output
> * Merge branch 'linux-system-roles:master' into add_user_and_password_ELK
> 
> The second commit is reverting the changes made on roles/rsyslog/tasks/set_certs.yml and tests/tests_files_elasticsearch.yml. Not sure what operation made it occur, but the second commit should not be there.
> 
> If you do `git log` in your branch `add_user_and_password_ELK`, the top part would look like this.
> 
> ```
> commit 687ddc95e91dc7a0a8ca7d4113dfc83544099b78 (HEAD -> pr/236, origin/pr/236)
> Merge: 59e7de1 5742b01
> Author: avlitman <64130977+avlitman@users.noreply.github.com>
> Date:   Tue Aug 31 20:18:34 2021 +0300
> 
>     Merge branch 'linux-system-roles:master' into add_user_and_password_ELK
> 
> commit 59e7de1e975e907d163dbd1586135f1618b2234d (tmp)
> Author: Aviv Litman <alitman@redhat.com>
> Date:   Thu Aug 5 16:24:51 2021 +0300
> 
>     Add user and password in elasticsearch output
> ```
> 
> Could you delete the top commit and push it as follows?
> 
> ```
> git reset --hard 59e7de1e975e907d163dbd1586135f1618b2234d
> git push --force avlitman add_user_and_password_ELK
> ```

Done

nhosoi commented:
@avlitman, thanks a lot. To verify the tox tests are fixed by @richm's merge, could you do this one more time?
```
git fetch --all
git rebase master
git push --force avlitman add_user_and_password_ELK
```
It should make all the CI tests pass (fingers crossed).

nhosoi commented:
[citest commit:469295bb0b5c4e7399a7bfcec2a855e0f908ce9f]

nhosoi commented:
[citest commit:59e7de1e975e907d163dbd1586135f1618b2234d]

avlitman commented:
> @avlitman, thanks a lot. To verify the tox tests are fixed by @richm's merge, could you do this one more time?
> 
> ```
> git checkout master
> git pull
> git checkout add_user_and_password_ELK
> git rebase master
> git push --force avlitman add_user_and_password_ELK
> ```
> 
> It should make all the CI tests pass (fingers crossed).

The problem is that everything is already up to date when trying to push:
[alitman@alitman logging]$ git checkout master
Switched to branch 'master'
Your branch is up to date with 'origin/master'.

[alitman@alitman logging]$ git pull
.
.
.
 + 7fa304e...59e7de1 add_user_and_password_ELK -> origin/add_user_and_password_ELK  (forced update)
Already up to date.

[alitman@alitman logging]$ git checkout add_user_and_password_ELK
Switched to branch 'add_user_and_password_ELK'

[alitman@alitman logging]$ git rebase master
Current branch add_user_and_password_ELK is up to date.

[alitman@alitman logging]$ git push --force avlitman add_user_and_password_ELK
.
.
Everything up-to-date

nhosoi commented:
> I was able to merge master into add_user_and_password_ELK and run:

No, please do not do that. If you do it, it makes this pr not mergeable to the master branch. :( Please undo the merge master part and push your add_user_and_password_ELK again. Thanks!

avlitman commented:
> > I was able to merge master into add_user_and_password_ELK and run:
> 
> No, please do not do that. If you do it, it makes this pr not mergeable to the master branch. :( Please undo the merge master part and push your add_user_and_password_ELK again (I think I fixed a mistake I could have made, and I'd like to see the result). Thanks!

Please undo the merge master part and push your add_user_and_password_ELK again,
You know How can I do that ?

nhosoi commented:
Well, if you like, please go ahead and do so...

As long as I observe, your branch add_user_and_password_ELK does not have any extra commits from master. I still can do the follows.
```
git checkout master
git pull
git checkout add_user_and_password_ELK
git rebase master
```
But if your local branch is mysteriously transformed, that's beyond my imagination... (;_;)

avlitman commented:
> @avlitman, thanks a lot. To verify the tox tests are fixed by @richm's merge, could you do this one more time?
> 
> ```
> git checkout master
> git pull
> git checkout add_user_and_password_ELK
> git rebase master
> git push --force avlitman add_user_and_password_ELK
> ```
> 
> It should make all the CI tests pass (fingers crossed).

Thanks, tried that before merging the master:

[alitman@alitman logging]$ git checkout master
Switched to branch 'master'
Your branch is up to date with 'origin/master'.

[alitman@alitman logging]$ git pull
warning: Pulling without specifying how to reconcile divergent branches is
discouraged. You can squelch this message by running one of the following
commands sometime before your next pull:

  git config pull.rebase false  # merge (the default strategy)
  git config pull.rebase true   # rebase
  git config pull.ff only       # fast-forward only

You can replace "git config" with "git config --global" to set a default
preference for all repositories. You can also pass --rebase, --no-rebase,
or --ff-only on the command line to override the configured default per
invocation.

Already up to date.

[alitman@alitman logging]$ git checkout add_user_and_password_ELK
Switched to branch 'add_user_and_password_ELK'

[alitman@alitman logging]$ git rebase master
Current branch add_user_and_password_ELK is up to date.

[alitman@alitman logging]$ git push --force avlitman add_user_and_password_ELK
Everything up-to-date

 

> Well, if you like, please go ahead and do so...
> 
> As long as I observe, your branch add_user_and_password_ELK does not have any extra commits from master. I still can do the follows.
> 
> ```
> git checkout master
> git pull
> git checkout add_user_and_password_ELK
> git rebase master
> ```
> 
> But if your local branch is mysteriously transformed, that's beyond my imagination... (;_;)

Thanks, tried that before merging the master:

[alitman@alitman logging]$ git checkout master
Switched to branch 'master'
Your branch is up to date with 'origin/master'.

[alitman@alitman logging]$ git pull
warning: Pulling without specifying how to reconcile divergent branches is
discouraged. You can squelch this message by running one of the following
commands sometime before your next pull:

  git config pull.rebase false  # merge (the default strategy)
  git config pull.rebase true   # rebase
  git config pull.ff only       # fast-forward only

You can replace "git config" with "git config --global" to set a default
preference for all repositories. You can also pass --rebase, --no-rebase,
or --ff-only on the command line to override the configured default per
invocation.

Already up to date.

[alitman@alitman logging]$ git checkout add_user_and_password_ELK
Switched to branch 'add_user_and_password_ELK'

[alitman@alitman logging]$ git rebase master
Current branch add_user_and_password_ELK is up to date.

[alitman@alitman logging]$ git push --force avlitman add_user_and_password_ELK
Everything up-to-date

avlitman commented:
@ I created a new pr: https://github.com/linux-system-roles/logging/pull/246 with all the changes, and compared several to make sure.
